File: python/showentr.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Simple command line tool to find and display entries
# in the JMdict database.

_VERSION_ = ("$Revision$"[11:-2], "$Date$"[7:-11])

import sys, os, re
import db, jbdb
from tables import *

class ParseError (RuntimeError): pass

global KW, Def_enc

def main (args, opts):
	global KW

	# open the database...
	try: cursor = db.dbOpen (user=opts.u, pw=opts.p, db=opts.d)
	except db.dbapi.OperationalError, e:
	    print "Error, unable to connect to database, do you need -u or -p?\n", str(e);  
	    sys.exit(1)

	# Read in all the keyword tables...
	KW = jbdb.Kwds (cursor)
	mk_tmpsrch (cursor)

	print """
Type "help" for more info.
To exit type a return, or EOF (usually ^d on Unix, ^z on Windows)."""

	while True:
	    try: s = raw_input( "find> ").decode(sys.stdout.encoding).strip()
	    except EOFError: break
	    if not s: break
	    if s[0] == "h" or s[0] == "H":
		help (); continue
	    try: rowcnt = search (cursor, s)
	    except ParseError, e: 
		print "Parse error, type \"h\" for help"; continue
	    if rowcnt == 0: print "Nothing found"
	    elif rowcnt == 1: 
		rs = get_found_list (cursor)
		entr = get_entry (cursor, rs[0][0])
		entr = display_entry (entr)
	    else:
		rs = get_found_list (cursor)
		choose_entry (cursor, rs)
	    
def search (cursor, s):
	Tbls = {'e':"entr",'k':"kanj",'r':"kana",'s':"sens",'g':"gloss",'q':"entr",None:"entr"}
	parsed = srch_parse (s)
	results = []
	for typ,tbl,val in parsed:
	    table = Tbls[tbl];  col = "txt"
	    if isinstance (val, (int,long)): col = "id"
	    if col=="txt" and table in ("entr","sens"):
		raise ParseError ("Parse error")
	    if typ is None: w = "%s.%s=%%s" % (table,col)
	    else: w = "%s.%s LIKE(%%s)" % (table,col)
	    if typ == '^': val = val + "%"
	    elif typ == '*': val = "%" + val + "%"
	    elif typ == '$': val = "%" + val
	    results.append ((table,w,(val,)))
	cursor.execute ("DELETE FROM _tmpsrch")
	sql, args = build_search_sql (results)
	cursor.execute ("INSERT INTO _tmpsrch(id) " + sql, args)
	return cursor.rowcount

def choose_entry (cursor, rs):
	display_list (rs)
	while True:
	    try: s = raw_input( "show> ").decode(sys.stdout.encoding).strip()
	    except EOFError: break
	    if not s: break
	    elif s[0] == "h" or s[0] == "H": help ()
	    elif s[0] == "l" or s[0] == "L": display_list (rs)
	    else:
		try: numb = int (s)
		except ValueError: 
		    print "Not a number or command, type \"h\" for help"; continue
		if numb < 1 or numb > len(rs):
		    print "Number out of range 1-%d, type \"h\" for help" % len(rs); continue
		else: 
		    entr = get_entry (cursor, rs[numb-1][0])
		    display_entry (entr)

def display_list (rs):
	print "Num.  eId     Seq             Readings |              Kanji | Glosses"
	for n,(e,q,k,j,g) in enumerate (rs):
	    u = "%3d.%6d%9d%s|%s|%s" % (n+1,e,q,clip(k,10,u"\u3000"),
			clip(j,10,u"\u3000"),clip(g,18," "))
	    # Explicitly convert to the system default excoding 
	    # for output, because we can get encoding errors which
	    # we don't want to bomb the program. 
	    s = u.encode (sys.stdout.encoding, "replace")
	    print s

def clip (s, n, pad):
	if isinstance(pad, unicode): s = s.replace("; ", u"\uff1b")
	x = min (len (s), n)
	s = (pad * (n - x)) + s[:x]
	return s

def get_found_list (cursor):
	sql = \
	    u"SELECT e.id,seq,kana,kanj,gloss " \
		"FROM entr_summary e JOIN _tmpsrch t ON e.id=t.id"
	cursor.execute (sql)
	rs = cursor.fetchall ()
	return rs

def display_entry (entr):

	  # Print basic entry information 

	print "\nseq: %s, lang=(%s), dial=(%s), id=%d\n" % (entr.seq, 
	    ",".join([KW.LANG[x.kw].descr for x in entr.lang]), 
	    ",".join([KW.DIAL[x.kw].descr for x in entr.dial]), entr.id)
	
	  # Print the entry's kanji text.
 
	klist = []
	for k in entr.kanj:
	    kwds = ";".join([KW.KINF[x.kw].kw for x in k.inf] \
		   + [KW.FREQ[x.kw].kw+str(x.value) for x in k.freq])
	    if kwds: klist.append (k.txt + "/" + kwds)
	    else: klist.append (k.txt)
	if klist: print "; ".join (klist)
	
	  # And the readings.  The "max(...) > 0" expression is 
	  # true if a restriction exists on any of the readings 
	  # in which case we call display_restr() instead of just 
	  # printing the readings.  We don't display r_inf or 
	  # r_rfreq for for simplicity -- method is same as for
	  # kanji.

	if max([len(r.restr) for r in entr.read]) > 0:
	    display_restr (entr.read, entr.kanj)
	else:
	    print ", ".join([x.txt for x in entr.read])

	  # Print the sense information.

	for n,s in enumerate (entr.sens):

	      # Part-of-speech, misc keywords, field...
	    pos = ";".join([KW.POS[p.kw].kw for p in s.pos])
	    if pos: pos = "[" + pos + "]"
	    misc = ";".join([KW.MISC[p.kw].kw for p in s.misc])
	    if misc: misc = "[" + misc + "]"
	    fld = ", ".join([KW.FLD[p.kw].descr for p in s.fld])

	      # Restrictions... 
	    if not s.stagr: sr = []
	    else: sr = filt ([x.kana for x in s.stagr], entr.read)
	    if not s.stagk: sk = []
	    else: sk = filt ([x.kanj for x in s.stagk], entr.kanj)
	    stag = ""
	    if sr or sk: 
		stag = " (%s only)" % ", ".join([x.txt for x in sk+sr])

	    print "\n  %d. %s%s%s(%d)" % (n+1, pos, misc, stag, s.id)
	    if fld: print "     Field(s): " + fld

	      # Now print the glosses...
	    for g in s.gloss:
		lang = KW.LANG[g.lang].kw
		if lang: lang += ": "
		  # Some glosses may contain characters not displayable
		  # in the user's system character set and which will 
		  # cause a UnicodeEncodeError if just printed.   So 
		  # explicitly encode with "replace" to avoid that.
		gtxt = g.txt.encode (sys.stdout.encoding, "replace")
	        print "     %s%s" % (lang, gtxt)

	      # Print the number of xrefs than this sense references,
	      # and the numbers of other senses that reference this
	      # one.

	    if len(s.xref) > 0: print "     [%d xrefs from this sense.]" % len(s.xref)
	    if len(s.xrex) > 0: print "     [%d xrefs to this sense.]" % len(s.xrex)

def display_restr (rlist, klist):
	s = []
	for r in rlist:
	    valid = filt ([x.kanj for x in r.restr], klist)
	    if not valid: s.append (r.txt)
	    else: s.append (r.txt + u" (" + ", ".join([j.txt for j in valid]) + u")")
	print "; ".join(s)

def get_entry (cursor, entrid):
	sql = "SELECT * FROM entr WHERE id=%s"
	e = load (Entr, cursor, sql, (entrid,))
	if len(e) < 1: return None
	e = e[0]; id = e.id
	e.lang = load (Lang, cursor, "select * from lang where entr=%s", (id,))
	e.dial = load (Dial, cursor, "select * from dial where entr=%s", (id,))
	e.read = load (Read, cursor, "select * from kana where entr=%s order by ord", (id,))
	e.kanj = load (Kanj, cursor, "select * from kanj where entr=%s order by ord", (id,))
	e.sens = load (Sens, cursor, "select * from sens where entr=%s order by ord", (id,))
	for s in e.sens:
	    s.gloss = load (Gloss, cursor, "select * from gloss where sens=%s order by ord", (s.id,))
	    s.pos = load (Pos, cursor, "select * from pos where sens=%s", (s.id,))
	    s.misc = load (Misc, cursor, "select * from misc where sens=%s", (s.id,))
	    s.fld = load (Fld, cursor, "select * from fld where sens=%s", (s.id,))
	    s.xref = load (Xref, cursor, "select * from xref where sens=%s", (s.id,))
	    s.xrex = load (Xref, cursor, "select * from xref where xref=%s", (s.id,))
	    s.stagr = load (Stagr, cursor, "select * from stagr where sens=%s", (s.id,))
	    s.stagk = load (Stagk, cursor, "select * from stagk where sens=%s", (s.id,))

	for j in e.kanj:
	    j.inf = load (Kinf, cursor, "select * from kinf where kanj=%s", (j.id,))
	    j.freq = load (Kfreq, cursor, "select * from kfreq where kanj=%s", (j.id,))
	for k in e.read:
	    k.inf = load (Kinf, cursor, "select * from rinf where kana=%s", (k.id,))
	    k.freq = load (Kfreq, cursor, "select * from rfreq where kana=%s", (k.id,))
	    k.restr = load (Restr, cursor, "select * from restr where kana=%s", (k.id,))
	return e

def build_search_sql (condlist):
	"""\
	Build a sql statement that will find the id numbers of
	all entries matching the conditions given in <condlist>.
	Note: This function does not provide for generating arbitrary 
	sql statements; it is only intented to support some limited 
	search capabilities.

	<condlist> is a list of 3-tuples.  Each 3-tuple specifies
	one condition:
	  0: Name of table, one of: "entr", "kana", "kanj", "sens", 
	    "gloss", "pos", "use".
	  1: Sql snippit that will be AND'd into the WHERE clause.
	    Field names must be qualified by table.  When looking 
	    for a value in a field.  A "?" may (and should) be used 
	    where possible to denote an exectime parameter.  The value
	    to be used when the sql is executed is is provided in
	    the 3rd member of the tuple (see #2 next).
	  2: A sequence of argument values for any exec-time parameters
	    ("?") used in the second value of the tuple (see #1 above).

	Example:
	    [("entr","entr.typ=1", ()),
	     ("gloss", "gloss.text LIKE ?", ("'%'+but+'%'",)),
	     ("pos","pos.kw IN (?,?,?)",(8,18,47))]

	  This will generate the SQL statement and arguments:
	    "SELECT entr.id FROM (((entr INNER JOIN sens ON sens.entr=entr.id) 
		INNER JOIN gloss ON gloss.sens=sens.id) 
		INNER JOIN pos ON pos.sens=sens.id) 
		WHERE entr.typ=1 AND (gloss.text=?) AND (pos IN (?,?,?))"
	    ('but',8,18,47)
	  which will find all entries that have a gloss containing the
	  substring "but" and a sense with a pos (part-of-speech) tagged
	  as a conjunction (pos.kw=8), a particle (18), or an irregular
	  verb (47)."""

	tables = []; wclauses = []; args =[]
	for tbl,cond,arg in condlist:
	    tables.append (tbl)
	    wclauses.append (cond)
	    args.extend (arg)
	t = "(%s INNER JOIN % s ON %s)"
	s = "entr "
	if "kana" in tables:  s = t % ( s, "kana",  "kana.entr=entr.id" )
	if "kanj" in tables:  s = t % ( s, "kanj",  "kanj.entr=entr.id" )
	if "sens" in tables \
		or "gloss" in tables \
		or "pos" in tables \
		or "misc" in tables:
			      s = t % ( s, "sens",  "sens.entr=entr.id" ) 
	if "gloss" in tables: s = t % ( s, "gloss", "gloss.sens=sens.id" ) 
	if "pos" in tables:   s = t % ( s, "pos",   "pos.sens=sens.id" ) 
	if "misc" in tables:  s = t % ( s, "misc",  "misc.sens=sens.id" ) 
	where = " AND ".join (wclauses)
	return "SELECT DISTINCT entr.id FROM %s WHERE %s" % (s, where), tuple(args)

def srch_parse (txt):
		  #   123          4         56       7      89           0  1
	regex = ur"\s*((([ekrsgq])?([0-9]+))|(([*^$])?([krg])(([^'\s]\S*)|(\'([^']*?)\'))))"
                  #   123        3 3      32 23     3 3     334         4 4  5      5  4321
	reob = re.compile (regex, re.I|re.U)
	start = 0;  results = []
	while start < len(txt):
	    m = reob.match (txt.strip(), start)
	    if not m: raise ParseError ("Parse error")
	    start = m.span()[1]
	    if m.group(4): results.append ((None,m.group(3),int(m.group(4))))
	    else: 
		if m.group(9): results.append ((m.group(6),m.group(7),m.group(9)))
		else: results.append ((m.group(6),m.group(7),m.group(11)))
	return results

def load (cls, cursor, sql, sqlargs):
	cursor.execute (sql, sqlargs)
	rs = cursor.fetchall ()
	return [cls (x) for x in rs]

def filt (exclude, xlist):
	# returns a list containing those items of <xlist>
	# that do not have an id number that is in the list
	# of numbers in <exclude>.
	return [x for x in xlist if x.id not in exclude]

def mk_tmpsrch (cursor):
	try: cursor.execute ("DROP TABLE _tmpsrch;")
	except db.dbapi.OperationalError, e:	# Mysql
	    if "Unknown table" not in e[1]: raise
	except db.dbapi.ProgrammingError, e:	# Postgresql
	    if "does not exist" not in str(e): raise
	    cursor.conn.rollback()
	if db.dbapi.__name__ == "psycopg2":
	    sql = \
	    "CREATE TABLE _tmpsrch (" \
		"id INT NOT NULL PRIMARY KEY, " \
		"ord SERIAL NOT NULL UNIQUE);"
	else:	# Assume MySQLdb
	    sql = \
	    "CREATE TABLE _tmpsrch (" \
		"id INT UNSIGNED NOT NULL PRIMARY KEY, " \
		"ord INT UNSIGNED NOT NULL AUTO_INCREMENT UNIQUE);"
	cursor.execute (sql)

def help ():
	print """\
At the "find>" prompt, you can search for an entry by
the id number of any of it's components, or by searching
for text strings.

To find an entry by id number, enter a single letter
indicating the table, followed by the id number.  The
acceptable letters are "e" (entr.id), "r" (kana.id),
"k" (kanj.id), "s" (sens.id), "g" (gloss.id).  The
letter "q" denotes the entr.seq data.  Examples:

	# Show the entry that has entr.id = 33108
	find> e33108	

	# Show the entry that has reading kana.id = 10558
	find> r10558

	# Show the entry with seq = 1000240
	find> q1000240

To find entries whose reading, kanji, or gloss contains
a particular string, use this syntax:

	[srch-type][table][text]

[srch-type] is optional and may be one of "^", "*", "$".
If not present, [text] must exactly match a the text in 
the searched table.  "^" denotes that [text] need only
occur at the start of the text in the table.  "*" denotes
[text] occurs anywhere in the table text.  "$" denotes
[text] occurs at the end of the table text.

[table] indicates the table to be searched: "r" readings
(table kana), "k" kanji (table kanj), "g" glosses (table
gloss).  

[text] is the text to be searched for.  If is contains
spaces, it must be enclosed in single quotes ("'").

Multiple search conditions may be given; matching entries
must satisfy all the conditions.

Examples:

	# Find all entries having a reading of "atsui"
	find> rあつい

	# Find all entries with kanji that starts with
	# "omo.u" and has a gloss "to feel"
	find> ^k思 g'to feel'

	# Find all entries with kanji that starts with
	# "omo.u" and has a gloss containing the word
	# "feel"
	find> ^k思 *gfeel

If only one entry is found, it is displayed.  If more than
entry is found, a numbered list is displayed, and you are
prompted with the "show>" prompt to enter the number of the
entry you want.  At the "show>" prompt you can also entr
"l" (lower case "L") to redisplay the list, "h" for this
help, or an blank line to return to the "find>" prompt.
""" 

#---------------------------------------------------------------------

from optparse import OptionParser

def parse_cmdline ():
	u = \
"""\n\t%prog [-d database][-u user][-p passwd] 

  %prog is a simple-minded program to illustrate extracting
  and displaying data from the jb database.  Options can be
  used to give information necessary for connecting to the
  database.  The program runs in a loop, prompting for an
  entry id or sequence number number, and displaying that
  entry.

  To exit, type a returnm or enter ^Z on MS Windows or ^D
  on Unix. 

arguments:  None"""

	v = sys.argv[0][max (0,sys.argv[0].rfind('\\')+1):] \
	        + " Rev %s (%s)" % _VERSION_
	p = OptionParser (usage=u, version=v)
	p.add_option ("-d", "--database",
             type="str", dest="d", default="jb",
             help="Name of the database to load.  Default is \"jb\".")
	p.add_option ("-u", "--user",
             type="str", dest="u", default="root", 
             help="Connect to the database with this username.  " \
                      "Deafult is \"root\"")
	p.add_option ("-p", "--passwd",
             type="str", dest="p", default="",
             help="Connect to that database with this password.")
	p.add_option ("-D", "--debug",
             action="store_true", dest="D", 
             help="Startup in Python pdb debugger.")
	opts, args = p.parse_args ()
	if len(args) > 0: p.error("Error, no arguments expected.  " 
				"Use --help for more info")
	return args, opts

def exceptionHandler (type, value, tb):
	if not (sys.stderr.isatty() and sys.stdin.isatty()) or type==SyntaxError: 
	    sys.__excepthook__(type, value, tb)
	else: # Not in interactive mode, print the exception...
	    import traceback, pdb
	    traceback.print_exception(type, value, tb)
	    print;  pdb.pm()

if __name__ == '__main__': 
	args, opts = parse_cmdline ()
	if (opts.D):
	    import pdb, traceback
	    sys.excepthook = exceptionHandler
	    pdb.set_trace ()
	main (args, opts)

File: perl/showentr.pl
#!/usr/bin/env perl
#######################################################################
#   This file is part of JMdictDB. 
#   Copyright (c) 2006,2007 Stuart McGraw 
# 
#   JMdictDB is free software; you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published 
#   by the Free Software Foundation; either version 2 of the License, 
#   or (at your option) any later version.
# 
#   JMdictDB is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
# 
#   You should have received a copy of the GNU General Public License
#   along with JMdictDB; if not, write to the Free Software Foundation,
#   51 Franklin Street, Fifth Floor, Boston, MA  02110#1301, USA
#######################################################################

@VERSION = (substr('$Revision$',11,-2), \
	    substr('$Date$',7,-11));

# to do:
# temp table very slow, use only for sense?
# fix encoding issues on windows.

use strict;  use warnings;
use Encode;  use DBI;
use Getopt::Std ('getopts');


BEGIN {push (@INC, "./lib");}
use jmdict;  use jmdictfmt;
$::Debug = {};

#-----------------------------------------------------------------------

    main: {
	my ($dbh, $entries, $e, @qlist, @elist, $enc, $host,
	    $dbname, $user, $pw, $typ, $corp, $numb);

	  # Read and parse command line options.
	if (!getopts ("hvd:u:p:r:e:j", \%::Opts) or $::Opts{h}) { usage (0); }

	  # Set some local variables based on the command line 
	  # options given or defaults where options not given.
	$enc =    $::Opts{e} || "utf8";
	$user =   $::Opts{u} || "postgres";
	$pw =     $::Opts{p} || "";
	$dbname = $::Opts{d} || "jmdict";
	$host =   $::Opts{r} || "";
	if ($::Opts{v}) { $::Debug{prtsql} = 1; }

	  # Make stderr unbuffered.
	my $oldfh = select(STDERR); $| = 1; select($oldfh);

	  # Set the default encoding of stdout and stderr.
	binmode(STDOUT, ":encoding($enc)");
	binmode(STDERR, ":encoding($enc)");

	  # Debugger writes to $DB::OUT.  Set its encoding but it
	  # will die if debugger not active.  So do inside and eval
	  # so as to not tereminate program in this case.
	eval { binmode($DB::OUT, ":encoding($enc)"); }; $dbh=$DB::OUT;

	  # Connect to the database.  Option PrintWarn is off to reduce
	  # message noise.  RaiseError is on so we don't need to check
	  # return code after every database operation,; error will
	  # cause exception.  AutoCommit off because we want to control
	  # transactions ourself.  (Although that is moot is this app 
	  # since we will only read from the database.)
	if ($host) { $host = ";host=$host"; }
	$dbh = DBI->connect("dbi:Pg:dbname=$dbname$host", $user, $pw, 
			{ PrintWarn=>0, RaiseError=>1, AutoCommit=>0 } );

	  # This is needed so that postgresql will give us unicode
	  # characters instead of bytes. 
	$dbh->{pg_enable_utf8} = 1;

	  # Read all the kw* tables put the data into a hash
	  # structure that we will use to conver id numbers to 
	  # keywords and visa-versa.  Save as a global variable
	  # so that data (which we will treat as aread-only)
	  # will be available thoughout this app.
	$::KW = Kwds ($dbh);

	  # Parse the command line arguments.
	foreach (@ARGV) {

	      # Build two lists from the command line arguments.
	      # @qlist will hold seq. numbers, @elist will hold
	      # entry id numbers.

	      # Check the command line argument format with a regex.
	    if (!(m/^([qe])?([0-9]+)(\.(\S+))?$/))  { 

		  # If no match, it has a bad format.
		print STDERR "Bad entry specifier in: $_\n";  next; }

	      # Get the three individual pieces of the argument format,
	      # providing default values for missing pieces. 
	    ($typ, $numb, $corp) = ($1 || 'q', $2, $4 || "jmdict");

	      # If it is an entry id specifier, push the value on to
	      # the @elist.

	    if ($typ eq 'e') { push (@elist, $numb); }

	      # Else (we know it is 'q' or otherwise the regex would
	      # have rejected it), check the the corpus values is 
	      # legit.  Note that one can lookup both keyword id numbers
	      # or the kw text in the $::KW structure.  Save the 'q'
	      # value and corpus id number on the @qlist. 
	    else {
		if (!($corp = $::KW->{SRC}{$corp}{id})) {
		    print STDERR "Invalid corpus value in: $_\n"; next; } 
		push (@qlist, [$numb, $corp]); } }

	  # Call get_entries() to read the desired entries from the
	  # database and construct runtime objects for them.
	$entries = get_entries ($dbh, \@elist, \@qlist);

	  # Go through the list of entries and print each.
	foreach $e (@$entries) { 
	    if ($::Opts{j}) { print jel_entr ($e); }
	    else { print fmt_entr ($e); } }

	print "--\nObj retrieval time: " . $::Debug->{'Obj retrieval time'} . "\n";
	print "Obj build time: " . $::Debug->{'Obj build time'} . "\n";
	print "Xrefsum retrieval time: " . $::Debug->{'Xrefsum retrieval time'} . "\n";

	  # Cleanly disconnect from the database (and thus avoid
	  # a warning message.)
	$dbh->disconnect(); }

#-----------------------------------------------------------------------

    sub get_entries { my ($dbh, $elist, $qlist) = @_;
	# $dbh -- Database handle for an open connection to jmdict db.
	# @$elist -- List of entry id numbers of entries to get.
	# @$qlist -- List of seq. numbers of entries to get.

	my (@whr, $sql, $entries);

	  # To retrieve the objects, we need a sql statement that 
	  # will return their id numbers.  The WHERE clause of that
	  # statement will be like
	  #   "WHERE e.id IN (e1,e2,e3,...) 
	  #        OR (src=c1 AND seq=q1)
	  #        OR (src=c2 AND seq=q2)
	  #        OR ..."
	  # (where e1,etc are the entry id numbers and q1,c1,etc are 
	  # the entry seq and corpus numbers of the entries we want.)

	if (@$qlist) { push (@whr, join(" OR ",
			map("(src=$_->[1] AND seq=$_->[0])",@$qlist))); }
	if (@$elist) { push (@whr, "e.id IN (" . join(",",map($_,@$elist)) . ")"); }

	  # Now we can create the sql statement, including creating 
	  # the WHERE clause by joining the twp pieces together with 
	  # a " OR ".  The join will not include the "or" if there 
	  # is only one piece (because there were only "q" numbers or 
	  # only "e" numbers).
	$sql = "SELECT e.id FROM entr e WHERE " . join (" OR ", @whr);

	  # Now we can give the sql statement to  EntrList() which  
	  # will use it to read the data for all the entries, contruct 
	  # objects for them, and return the list of objects to us...
	$entries = EntrList ($dbh, $sql);

	  # The entries returned by EntrList() contain only the raw
	  # xref records (entry and sense numbers) which are not very 
	  # useful for display.  Call add_xrefsums() to get summary
	  # info for all those xrefs.  By giving the third arg, we ask
	  # add_xrefsums to distribute the summary info into each xref,
	  # in key {ssum}.
	add_xrefsums ($dbh, $entries);

	  # ... which we return to our caller.
	return $entries; }

#-----------------------------------------------------------------------

sub usage { my ($exitstat) = @_;
	print <<EOT;

Usage: showentr.pl [options] ['e'entry_id] [['q']entry_seq[.[n|corpus]]] ...

    Arguments:
	A list of entries to display where each entry is specified
	in one of the following formats:
	  * nnnn -- Entry sequence number <nnnn>.
	  * ennnn -- Entry id number <nnnn>.
	  * qnnnn -- Entry sequence number <nnnn> in corpus "jmdict".
	  * qnnnn.mm -- Entry sequence number <nnnn> in corpus
	      number <mm>.
	  * qnnnn.aaaaa -- Entry sequence number in corpus named 
	      <aaaaa>, for example, "q14359.jmnedict".

    Options:
	-d dbname -- Name of database to use.  Default is "jmdict".
	-r host	-- Name of machine hosting the database.  Default
		is "localhost".
	-e encoding -- Encoding to use for stdout and stderr.
	 	Default is "utf-8".  Windows users running with 
		a Japanese locale may wish to use "cp932".
	-j -- Generate output in "JEL" format.  
	-v -- (Verbose) print debugging info.
	-h -- (help) print this text and exit.

	  ***WARNING***
	  The following two options are not recommended because 
	  their values will be visible to anyone who can run a 
	  \"ps\" command.

	-u username -- Username to use when connecting to database.
	        Default is "postgres".
	-p password -- Password to use when connecting to database.
	        No default.
EOT
	exit $exitstat; }

	

	 